include/unifex/task.hpp

```diff
@@ -290,27 +306,56 @@ struct _awaiter {
       auto thisCoro = coro::coroutine_handle<ThisPromise>::from_address((void*) coro_);
       ++coro_; // mark the awaiter as needing cleanup
       auto& promise = thisCoro.promise();
-      sched_.construct(get_scheduler(h.promise()));
       promise.continuation_ = h;
-      promise.stoken_ = stopTokenAdapter_.subscribe(get_stop_token(h.promise()));
-      promise.sched_ = sched_.get();
+      if constexpr (needs_scheduler_t::value) {
+        sched_.construct(get_scheduler(h.promise()));
+        promise.sched_ = sched_.get();
+      } else {
+        promise.sched_ = get_scheduler(h.promise());
+      }
+      if constexpr (needs_stop_token_t::value) {
+        promise.stoken_ = stopTokenAdapter_.subscribe(get_stop_token(h.promise()));
+      } else {
+        promise.stoken_ = get_stop_token(h.promise());
+      }
       return thisCoro;
     }
 
     result_type await_resume() {
-      stopTokenAdapter_.unsubscribe();
-      sched_.destruct();
+      if constexpr (needs_stop_token_t::value)
+        stopTokenAdapter_.unsubscribe();
+      if constexpr (needs_scheduler_t::value)
+        sched_.destruct();
       auto thisCoro = coro::coroutine_handle<ThisPromise>::from_address(
           (void*) std::exchange(--coro_, 0));
       scope_guard destroyOnExit{[&]() noexcept { thisCoro.destroy(); }};
       return thisCoro.promise().result();
     }
 
   private:
+    using scheduler_t = get_scheduler_result_t<OtherPromise&>;
+    using stop_token_t = stop_token_type_t<OtherPromise>;
+    using needs_scheduler_t =
+        std::bool_constant<!same_as<scheduler_t, any_scheduler_ref>>;
+    using needs_stop_token_t =
+        std::bool_constant<!same_as<stop_token_t, inplace_stop_token>>;
```

Comment:
Should these be `remove_cvref_t`ed?
```suggestion
        std::bool_constant<!same_as<remove_cvref_t<scheduler_t>, any_scheduler_ref>>;
    using needs_stop_token_t =
        std::bool_constant<!same_as<remove_cvref_t<stop_token_t>, inplace_stop_token>>;
```